tests/robustness/model/replay.go

```diff
@@ -62,45 +63,74 @@ func (r *EtcdReplay) EventsForWatch(watch WatchRequest) (events []PersistedEvent
 	return events
 }
 
-func toWatchEvents(prevState *EtcdState, request EtcdRequest, response MaybeEtcdResponse) (events []PersistedEvent) {
-	if request.Type != Txn || response.Error != "" {
+func toWatchEvents(prevState *EtcdState, newState *EtcdState, request EtcdRequest, response MaybeEtcdResponse) (events []PersistedEvent) {
+	if response.Error != "" {
 		return events
 	}
-	var ops []EtcdOperation
-	if response.Txn.Failure {
-		ops = request.Txn.OperationsOnFailure
-	} else {
-		ops = request.Txn.OperationsOnSuccess
-	}
-	for _, op := range ops {
-		switch op.Type {
-		case RangeOperation:
-		case DeleteOperation:
-			e := PersistedEvent{
-				Event: Event{
-					Type: op.Type,
-					Key:  op.Delete.Key,
-				},
-				Revision: response.Revision,
-			}
-			if _, ok := prevState.KeyValues[op.Delete.Key]; ok {
+
+	switch request.Type {
+	case Txn:
+		var ops []EtcdOperation
+		if response.Txn.Failure {
+			ops = request.Txn.OperationsOnFailure
+		} else {
+			ops = request.Txn.OperationsOnSuccess
+		}
+		for _, op := range ops {
+			switch op.Type {
+			case RangeOperation:
+			case DeleteOperation:
+				e := PersistedEvent{
+					Event: Event{
+						Type: op.Type,
+						Key:  op.Delete.Key,
+					},
+					Revision: response.Revision,
+				}
+				if _, ok := prevState.KeyValues[op.Delete.Key]; ok {
+					events = append(events, e)
+				}
+			case PutOperation:
+				_, leaseExists := prevState.Leases[op.Put.LeaseID]
+				if op.Put.LeaseID != 0 && !leaseExists {
+					break
+				}
+
+				e := PersistedEvent{
+					Event: Event{
+						Type:  op.Type,
+						Key:   op.Put.Key,
+						Value: op.Put.Value,
+					},
+					Revision: response.Revision,
+				}
+				if _, ok := prevState.KeyValues[op.Put.Key]; !ok {
+					e.IsCreate = true
+				}
 				events = append(events, e)
+			default:
+				panic(fmt.Sprintf("unsupported operation type: %v", op))
+			}
+		}
+	case LeaseRevoke:
+		deletedKeys := []string{}
+		for key := range prevState.KeyValues {
+			if _, ok := newState.KeyValues[key]; ok {
```

Comment:
QQ, why not check keys that are connected to the lease?

Reply:
Correct! I think we should copy that whole EtcdState instead of only KeyValues. So, we can rebuild the deleted keys by prevState.

Reply:
Oh, I see. You mean that the model `Step` functions doesn't copy the `KeyLeases` dict like `KeyValues` map. Right that's a bug.  Overall I would like to rewrite the `KeyLeases` and `Leases` maps and combine them into one.